File: LowestCommonAncestorBST_iterative.py
# TC: O(logN) in average case, since we would be skipping half side of tree if p and q are greater or lesser than the root and O(N) in worst case. 
# SC: O(1) since we do not use any extra space.

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        while True: 
            if p.val < root.val and q.val < root.val: 
                root = root.left
                
            elif p.val > root.val and q.val > root.val: 
                root = root.right
            else: 
                return root
        

File: Lowest-common-ancestor-bst-iterative.java
// Time Complexity : O(N)

// Space Complexity : O(1)

// Did this code successfully run on Leetcode : YES

// Appoarch: Using dfs, if root is greater than p and q, move left else
// move right or return root if p is in left and q is in the right. 

// 235. Lowest Common Ancestor of a Binary Search Tree

// Iterative

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // Iterative
        while(true){
            if(root.val > p.val && root.val > q.val){
            root = root.left;
        }
        else if(root.val < p.val && root.val < q.val){
            root = root.right;
        }
        else{
            return root;
        }
        }  
    }
}